File: src/main/scala/net/projecteuler/scala/Problem145.scala
package net.projecteuler.scala

import scala.collection.mutable.BitSet

/**
 * Some positive integers n have the property that the sum [ n + reverse(n) ]
 * consists entirely of odd (decimal) digits. For instance, 36 + 63 = 99
 * and 409 + 904 = 1313. We will call such numbers <i>reversible</i>; so 36, 63, 409,
 * and 904 are reversible. Leading zeroes are not allowed in either n or reverse(n).</br>
 * There are 120 reversible numbers below one-thousand.</br>
 * How many reversible numbers are there below one-billion (10<sup>9</sup>)?
 *
 * @author vbox
 *
 */
object Problem145 {

    def main(args: Array[String]): Unit = {
        // There are 9 digit solution, because of leading zero - we calculate till 10^8
        val limit = 100000000
        // BitSet is used to mark the number and its reversed one in order to avoid
        // calculation twice, because of performance
        val numbers = new BitSet(limit)
        for (i <- 0 to limit by 10) {
            numbers.update(i, true)
        }
        var counter = 0
        for (i <- 1 to limit) {
            if (!numbers.contains(i)) {
                val reversedNumber = reverse(i, 0)
                // Sets the reversed value in the bit set - do not calculate that value twice
                numbers.update(reversedNumber, true)
                if (reversible(i + reversedNumber)) {
                    counter += 1
                }
            }
        }

        assert(counter * 2 == 608720)
    }

    /**
     * Produces the reverse of the given number using of tail recursion
     */
    def reverse(number: Int, reversedNumber: Int): Int = {
        if (number == 0)
            reversedNumber
        else
            reverse(number / 10, reversedNumber * 10 + number % 10)
    }

    /**
     * Check the given number reversible or not based on the sum.
     */
    def reversible(sum: Int): Boolean = {
        if (sum == 0)
            true
        else if (sum % 2 == 0) {
            false
        } else {
            reversible(sum / 10)
        }
    }
}

File: src/main/java/net/projecteuler/java/Problem145.java
package net.projecteuler.java;

import java.util.BitSet;

/**
 * Some positive integers n have the property that the sum [ n + reverse(n) ]
 * consists entirely of odd (decimal) digits. For instance, 36 + 63 = 99
 * and 409 + 904 = 1313. We will call such numbers <i>reversible</i>; so 36, 63, 409,
 * and 904 are reversible. Leading zeroes are not allowed in either n or reverse(n).</br>
 * There are 120 reversible numbers below one-thousand.</br>
 * How many reversible numbers are there below one-billion (10<sup>9</sup>)?
 *
 * @author vbox
 *
 */
public class Problem145 {

	public static void main(String... args) {
		assert solve() == 608720 : "failed";
	}

	private static int solve() {
		int limit = 100000000; // There are 9 digit solution, because of leading zero - we calculate till 10^8
		// BitSet is used to mark the number and its reversed one in order to avoid
		// calculation twice, because of performance
		BitSet numbers = new BitSet(limit);

		for (int i = 0; i < numbers.size(); i = i + 10) {
			numbers.set(i, true); // We do not calculate with multiple of 10
		}
		int counter = 0;
		for (int i = 1; i < limit; i++) {
			if (numbers.get(i)) {
				continue;
			}
			int reversedNum = 0, number = i;
			while (number != 0) {
				reversedNum = reversedNum * 10 + number % 10;
				number = number / 10;
			}

			// Sets the reversed value in the bit set
			numbers.set(reversedNum, true);

			int sum = i + reversedNum;
			boolean passed = true;
			while (sum != 0) {
				if (sum % 2 == 0) {
					passed = false;
					break;
				}
				sum = sum / 10;
			}
			if (passed) {
				++counter;
			}
		}

		return counter * 2;
	}
}